src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -439,13 +439,36 @@ sliceSepBy _cond = undefined
 
 -- | See 'Streamly.Internal.Data.Parser.sliceBeginWith'.
 --
--- /Unimplemented/
+-- /Internal/
 --
-{-# INLINABLE sliceBeginWith #-}
-sliceBeginWith ::
-    -- Monad m =>
-    (a -> Bool) -> Fold m a b -> Parser m a b
-sliceBeginWith = undefined
+data SliceBeginWithState s = Left' s | Right' s
+
+{-# INLINE sliceBeginWith #-}
+sliceBeginWith :: Monad m => (a -> Bool) -> Fold m a b -> Parser m a b
+sliceBeginWith cond (Fold fstep finitial fextract) =
+
+    Parser step initial extract
+
+    where
+
+    initial = Left' <$> finitial
+
+    {-# INLINE process #-}
+    process s a = do
+        res <- fstep s a
+        return
+            $ case res of
+                FL.Partial s1 -> Partial 0 (Right' s1)
+                FL.Done b -> Done 0 b
+
+    step (Left' s) a = process s a
```

Comment:
We should throw an error if the slice does not begin with the expected element. If we want a parser to be lax such as to allow the first segment to not begin with the element then we can compose it with a `takeWhile` parser.